arch/x86/core/thread.c

```diff
@@ -175,93 +106,85 @@ __asm__("\t.globl _thread_entry\n"
+ * for when _Swap() switches to it for the first time.
+ */
+struct _x86_initial_frame {
+	u32_t swap_retval;
+	u32_t ebp;
+	u32_t ebx;
+	u32_t esi;
+	u32_t edi;
+	void *_thread_entry;
+	u32_t eflags;
+	_thread_entry_t entry;
+	void *p1;
+	void *p2;
+	void *p3;
+};
+
 /**
- *
  * @brief Create a new kernel execution thread
  *
- * This function is utilized to create execution threads for both fiber
- * threads and kernel tasks.
+ * Initializes the k_thread object and sets up initial stack frame.
  *
  * @param thread pointer to thread struct memory, including any space needed
  *		for extra coprocessor context
- * @param pStackmem the pointer to aligned stack memory
- * @param stackSize the stack size in bytes
- * @param pEntry thread entry point routine
+ * @param stack the pointer to aligned stack memory
+ * @param stack_size the stack size in bytes
+ * @param entry thread entry point routine
  * @param parameter1 first param to entry point
  * @param parameter2 second param to entry point
  * @param parameter3 third param to entry point
  * @param priority thread priority
  * @param options thread options: K_ESSENTIAL, K_FP_REGS, K_SSE_REGS
- *
- *
- * @return opaque pointer to initialized k_thread structure
  */
 void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
-		 size_t stackSize,
-		 _thread_entry_t pEntry,
+		 size_t stack_size, _thread_entry_t entry,
 		 void *parameter1, void *parameter2, void *parameter3,
 		 int priority, unsigned int options)
 {
-	char *pStackMem;
-
-	_ASSERT_VALID_PRIO(priority, pEntry);
-
-	unsigned long *pInitialThread;
+	char *stack_buf;
+	char *initial_esp;
+	struct _x86_initial_frame *initial_frame;
 
+	_ASSERT_VALID_PRIO(priority, entry);
 #if CONFIG_X86_STACK_PROTECTION
 	_x86_mmu_set_flags(stack, MMU_PAGE_SIZE, MMU_ENTRY_NOT_PRESENT,
 			   MMU_PTE_P_MASK);
 #endif
-	pStackMem = K_THREAD_STACK_BUFFER(stack);
-	_new_thread_init(thread, pStackMem, stackSize, priority, options);
-
-	/* carve the thread entry struct from the "base" of the stack */
-
-	pInitialThread =
-		(unsigned long *)STACK_ROUND_DOWN(pStackMem + stackSize);
-
-	/*
-	 * Create an initial context on the stack expected by the _Swap()
-	 * primitive.
-	 */
-
-	/* push arguments required by _thread_entry() */
-
-	*--pInitialThread = (unsigned long)parameter3;
-	*--pInitialThread = (unsigned long)parameter2;
-	*--pInitialThread = (unsigned long)parameter1;
-	*--pInitialThread = (unsigned long)pEntry;
-
-	/* push initial EFLAGS; only modify IF and IOPL bits */
-
-	*--pInitialThread = (EflagsGet() & ~EFLAGS_MASK) | EFLAGS_INITIAL;
-
+	stack_buf = K_THREAD_STACK_BUFFER(stack);
+	_new_thread_init(thread, stack_buf, stack_size, priority, options);
+
+	initial_esp = (char *)STACK_ROUND_DOWN(stack_buf + stack_size);
+
+	/* Create an initial context on the stack expected by _Swap() */
+	initial_frame = (struct _x86_initial_frame *)
+		(initial_esp - sizeof(struct _x86_initial_frame));
```

Comment:
Style: naming glitch: "initial_esp" indicates that this will go into the %ESP register, but it obviously can't, because here we're putting stuff *below* it, which is illegal the second that stack takes an interrupt.  Maybe call this thing "stack_top" or somesuch?

Reply:
yeah I'll rename this.